src/Generic/Data/Internal/Error.hs

```diff
@@ -0,0 +1,33 @@
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Generic.Data.Internal.Error where
+
+import Data.Type.Bool
+import GHC.Generics
+import GHC.TypeLits
+
+type family HasSum f where
+  HasSum V1 = 'False
+  HasSum U1 = 'False
+  HasSum (K1 i c) = 'False
+  HasSum (M1 i c f) = HasSum f
+  HasSum (f :*: g) = HasSum f || HasSum g
+  HasSum (f :+: g) = 'True
+
+type family AssertH pred msg where
+  AssertH 'True msg = '()
+  AssertH 'False msg = TypeError msg
+
+class Assert (pred :: Bool) (msg :: ErrorMessage)
+instance AssertH pred msg ~ '() => Assert pred msg
+
+type SemigroupAssertNoSum a =
+    Assert (Not (HasSum (Rep a)))
+    ('Text "No generic instances of " ':<>: 'ShowType Semigroup ':<>: 'Text " for sum types")
```

Comment:
Parameterizing this by the class would let us not worry about it being in scope here for old versions of base, whereas it will be at the use site(s). It might also be useful to mention the type `a` in this error message.

Reply:
Done